worker/firewaller/firewaller.go

```diff
@@ -1326,22 +1282,32 @@ func (fw *Firewaller) startRelation(rel *params.RemoteRelation, role charm.Relat
 		endpointRole:        role,
 		relationReady:       make(chan remoteRelationInfo),
 	}
-	fw.relationIngress[tag] = data
 
-	err = catacomb.Invoke(catacomb.Plan{
-		Site: &data.catacomb,
-		Work: data.watchLoop,
-	})
-	if err != nil {
-		delete(fw.relationIngress, tag)
+	// Start the relation worker and wait for it to succeed or fail startup.
+	// Under normal circumstances, unless there's a coding error, catacomb.Invoke
+	// should not return an error and thus the worker run loop should successfully
+	// start. Nonetheless, we'll be cautious and wait for the worker to start and if
+	// there's an error, exit.
+	var workerErr error
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+	if err := fw.relationWorkerRunner.StartWorker(tag.Id(), func() (worker.Worker, error) {
+		defer wg.Done()
```

Comment:
If the function is called twice, because the worker exited with an error, the WaitGroup will become negative, and panic.
What's the point of waiting for it to start anyway? Whether it fails to initialise, or fails part way through - why should we distinguish?

Reply:
There'a a difference between the runner not starting / initialising the worker and the worker exiting with a business logic error once running. In the first case, I thought it best to surface that initialisation error in the main worker so that the whole lot restarts and things get another go at getting set up. Maybe that's overkill?

Reply:
Yeah, I think so. If it were to fail on initialisation, it's most likely going to always fail. I think it would be simpler and less error-prone to consider them the same.